docs/recipes/UsageWithTypescript.md

```diff
@@ -6,259 +6,295 @@ hide_title: true
 
-export function systemReducer(
-  state = initialState,
-  action: SystemActionTypes
-): SystemState {
-  switch (action.type) {
-    case UPDATE_SESSION: {
-      return {
-        ...state,
-        ...action.payload
-      }
-    }
-    default:
-      return state
-  }
-}
+```ts
+export interface Middleware<
+  DispatchExt = {}, // legacy type parameter that is no longer relevant
+  S = any, // type of the Redux store state
+  D extends Dispatch = Dispatch // type of the dispatch method
+>
 ```
 
-We now need to generate the root reducer function, which is normally done using `combineReducers`. Note that we do not have to explicitly declare a new interface for RootState. We can use `ReturnType` to infer state shape from the `rootReducer`.
+A custom middleware should use the `Middleware` type, and pass the generic args for `S` (state) and `D` (dispatch) if needed:
 
 ```ts
-// src/store/index.ts
-
-import { systemReducer } from './system/reducers'
-import { chatReducer } from './chat/reducers'
+import { Middleware } from 'redux'
 
-const rootReducer = combineReducers({
-  system: systemReducer,
-  chat: chatReducer
-})
+import { RootState } from '../store'
 
-export type RootState = ReturnType<typeof rootReducer>
+export const exampleMiddleware: Middleware<
+  {}, // legacy type parameter added to satisfy interface signature
+  RootState
+> = storeApi => next => action => {
+  const state = storeApi.getState() // correctly typed as RootState
+}
 ```
 
-### Type Checking Middlewares
+The dispatch generic should likely only be needed if you are dispatching additional thunks within the middleware.
 
-A middleware is a higher-order function that composes a dispatch function to return a new dispatch function.
-To create a type checked middleware function, we can leverage the `Middleware` interface provided by Redux together with our store type (as defined in the previous example):
+### Type Checking Redux Thunks
+
+[Redux Thunk](https://github.com/reduxjs/redux-thunk) is the standard middleware for writing sync and async logic that interacts with the Redux store. A thunk function receives `dispatch` and `getState` as its parameters. Redux Thunk has a built in `ThunkAction` type which we can use to define types for those arguments:
 
 ```ts
-// src/middleware/example.ts
+export type ThunkAction<
+  R, // Return type of the thunk function
+  S, // state type used by getState
+  E, // any "extra argument" injected into the thunk
+  A extends Action // known types of actions that can be dispatched
+> = (dispatch: ThunkDispatch<S, E, A>, getState: () => S, extraArgument: E) => R
+```
 
-import { Middleware } from 'redux'
+You will typically want to provide the `R` (return type) and `S` (state) generic arguments. Unfortunately, TS does not allow only providing _some_ generic arguments, so the usual values for the other arguments are `unknown` for `E` and `Action<string>` for `A`:
```

Comment:
I'd personally prefer `AnyAction` over `Action<string>` because it seems to convey the intent a bit more, but that's just my nitpick.